From 136487aee5faa429a0caae758893850ad3b80cc9 Mon Sep 17 00:00:00 2001
From: Matthew Ballance <matt.ballance@gmail.com>
Date: Sun, 1 Feb 2026 20:24:29 +0000
Subject: [PATCH] build: add install-dev-sdk target for device module
 development

Add a new build option 'install_dev_sdk' that installs headers and
configuration files needed for building QEMU device modules as
external shared libraries.

This enables binary distributions to provide a device development SDK
without shipping the full source code.

The SDK includes:
- Public API headers from include/{qemu,qom,hw,exec,sysemu,chardev,io,migration,monitor}
- Generated configuration files (config-host.h)
- Generated QAPI headers
- pkg-config file for easy integration

Usage:
  meson configure -Dinstall_dev_sdk=true
  ninja install

Headers are installed to ${prefix}/include/qemu-device/ and can be
accessed via 'pkg-config --cflags qemu-device'.
---
 meson.build       | 66 +++++++++++++++++++++++++++++++++++++++++++++++
 meson_options.txt |  3 +++
 2 files changed, 69 insertions(+)

diff --git a/meson.build b/meson.build
index 147097c..6879dc7 100644
--- a/meson.build
+++ b/meson.build
@@ -4322,6 +4322,72 @@ if get_option('plugins')
   endif
 endif
 
+###########################
+# Device Module SDK       #
+###########################
+
+if get_option('install_dev_sdk')
+  # Install all public API headers needed for device development
+  install_subdir('include/qemu',
+                 install_dir: get_option('includedir') / 'qemu-device',
+                 strip_directory: false,
+                 exclude_files: ['osdep.h.inc'])
+
+  install_subdir('include/qom',
+                 install_dir: get_option('includedir') / 'qemu-device',
+                 strip_directory: false)
+
+  install_subdir('include/hw',
+                 install_dir: get_option('includedir') / 'qemu-device',
+                 strip_directory: false)
+
+  install_subdir('include/exec',
+                 install_dir: get_option('includedir') / 'qemu-device',
+                 strip_directory: false)
+
+  install_subdir('include/sysemu',
+                 install_dir: get_option('includedir') / 'qemu-device',
+                 strip_directory: false)
+
+  install_subdir('include/chardev',
+                 install_dir: get_option('includedir') / 'qemu-device',
+                 strip_directory: false)
+
+  install_subdir('include/io',
+                 install_dir: get_option('includedir') / 'qemu-device',
+                 strip_directory: false)
+
+  install_subdir('include/migration',
+                 install_dir: get_option('includedir') / 'qemu-device',
+                 strip_directory: false)
+
+  install_subdir('include/monitor',
+                 install_dir: get_option('includedir') / 'qemu-device',
+                 strip_directory: false)
+
+  # Install generated configuration files
+  install_data(meson.current_build_dir() / 'config-host.h',
+               install_dir: get_option('includedir') / 'qemu-device')
+
+  # Install generated QAPI headers (subdirectory with all files)
+  install_subdir(meson.current_build_dir() / 'qapi',
+                 install_dir: get_option('includedir') / 'qemu-device',
+                 strip_directory: false,
+                 exclude_directories: ['trace'])
+
+  # Generate and install pkg-config file
+  pkg = import('pkgconfig')
+  pkg.generate(
+    name: 'qemu-device',
+    description: 'QEMU Device Development SDK',
+    version: meson.project_version(),
+    variables: [
+      'includedir=${prefix}/include',
+    ],
+    subdirs: ['qemu-device'],
+  )
+endif
+
 subdir('qga')
 
 # Don't build qemu-keymap if xkbcommon is not explicitly enabled
diff --git a/meson_options.txt b/meson_options.txt
index 5eeaf3e..1f45f9a 100644
--- a/meson_options.txt
+++ b/meson_options.txt
@@ -382,3 +382,6 @@ option('rust', type: 'feature', value: 'disabled',
        description: 'Rust support')
 option('strict_rust_lints', type: 'boolean', value: false,
        description: 'Enable stricter set of Rust warnings')
+
+option('install_dev_sdk', type: 'boolean', value: true,
+       description: 'Install device development SDK (headers and configs)')
-- 
2.43.0

